### PR TITLE
feat: update new learner display and logic for improved visibility

### PR DESCRIPTION
**Pull request by naincy128**
### Description

This update introduces a new feature that displays a **"New Learner" message** beneath usernames in discussion posts and comments for users identified as new learners. The feature helps learners and staff quickly recognize new participants, improving clarity and community interaction.
 
The implementation integrates backend API fields for learner status, with fallback logic to support legacy and edge cases. Core discussion components were updated to pass learner status data through the component tree, along with new hooks and test coverage to ensure consistent and reliable functionality.
 
### Features

- Added a visible **“👋 Hi, I am a new learner”** message below usernames in discussion posts and comments for new learners, using backend API fields and fallback heuristics.  

- Added a **regular learner message** for users who no longer qualify as new learners.  

- Created new hooks:  

  - `useLearnerStatus` – encapsulates learner status logic and consumes API data.  

  - `useIsNewLearner` – determines whether a user qualifies as a new learner.  

- Updated core components including `AuthorLabel`, `CommentHeader`, `Reply`, and `Post` to receive and propagate a new `postData` prop containing learner status.  

- Updated prop types and default props across affected components to support the new prop.  

- Added **internationalization (i18n)** support for the new learner and learner messages in `messages.js`.  

- Added comprehensive **test coverage** to verify correct rendering behavior for different user roles (staff, moderators, anonymous, retired, regular, and new learners).
 
### Linked PRs

- **Frontend PR:** [edx/frontend-app-discussions](https://github.com/edx/frontend-app-discussions/pull/5)
 
### Screenshot
Screenshot 2025-10-27 105154

Screenshot 2025-10-27 105236

 
## Jira Ticket

[COSMO2-735](https://2u-internal.atlassian.net/jira/software/c/projects/COSMO2/boards/2821?selectedIssue=COSMO2-735)

### CODE REVIEW

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 1]
```diff
@@ -275,6 +277,58 @@ def get_author_label(self, obj):
...
             return self._get_user_label(user_id)
 
+    def get_is_new_learner(self, obj):
+        """
+        Returns whether the content author is a new learner.
+
+        A new learner is a user who has enrolled in the course but has not
+        yet engaged with any course content beyond viewing the course outline.
+        """
+        from .utils import is_user_new_learner
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: This is already imported at https://github.com/edx/edx-platform/blob/4120ba161ef892a7d6d97e2155d3e8e2eda9857b/lms/djangoapps/discussion/rest_api/serializers.py#L36-L40 

So you don't need to import it here.

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 2]
```diff
@@ -275,6 +277,58 @@ def get_author_label(self, obj):
...
+
+        # Return False for anonymous content
+        if self._is_anonymous(obj) or obj["user_id"] is None:
+            return False
+
+        # Get the user object
+        try:
+            user = User.objects.get(id=int(obj["user_id"]))
+        except (User.DoesNotExist, ValueError):
+            return False
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: This'd mean that if a user does not exist, it'll return is_new_learner = False ?

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 3]
```diff
@@ -275,6 +277,58 @@ def get_author_label(self, obj):
...
+        return is_user_new_learner(user, course.id)
+
+    def get_is_regular_learner(self, obj):
+        """
+        Returns whether the content author is a regular learner.
+
+        A regular learner is a user who has enrolled in the course and has
+        engaged with course content (no longer a new learner), but is not staff.
+        """
+        from .utils import is_user_regular_learner
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: Same as above, you can write this import at the top of the file where utils is already being imported.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 4]
```diff
@@ -496,3 +497,167 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+    Args:
+        user: User object to check
+        course_id: Course key to check engagement in
+
+    Returns:
+        bool: True if user is a new learner, False otherwise
+    """
+    from lms.djangoapps.courseware.models import StudentModule
+    from common.djangoapps.student.models import CourseEnrollment
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: These imports can be written at the top

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 5]
```diff
@@ -496,3 +497,167 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+    # Check if user has engaged with course content
+    # Look for StudentModule records that indicate meaningful engagement
+    engagement_modules = StudentModule.objects.filter(
+        student=user,
+        course_id=course_id
+    ).exclude(
+        # Exclude records that don't indicate real engagement
+        module_type='course'  # Course-level modules don't indicate content engagement
+    )
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: This is a very costly query to fire in terms of DB time, there may be existing functions that get the progress or completion of the course in some other method, try to go for that instead of this.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 6]
```diff
@@ -496,3 +497,167 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+    # Check for any meaningful engagement indicators
+    has_engagement = engagement_modules.filter(
+        # Look for any of these engagement indicators:
+        # 1. Grade is not null (user submitted something)
+        # 2. State is not empty (user interacted with content)
+        # 3. Done status indicates completion
+        models.Q(grade__isnull=False) |
+        models.Q(state__isnull=False, state__gt='') |
+        models.Q(done__in=['f', 'i'])  # 'f' = finished, 'i' = incomplete but started
+    ).exists()
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: Again, there might be better ways around this. StudenModule table is probably the bulkiest of the tables, so querying it directly is almost always not a very good idea. Try to see if there are existing methods which do what you want.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 7]
```diff
@@ -496,3 +497,167 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+    Args:
+        user: User object to check
+        course_id: Course key to check engagement in
+
+    Returns:
+        bool: True if user is a regular learner, False otherwise
+    """
+    from lms.djangoapps.courseware.models import StudentModule
+    from common.djangoapps.student.models import CourseEnrollment
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: Imports can go at the top

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 8]
```diff
@@ -496,3 +497,167 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+    # Check for any meaningful engagement indicators
+    has_engagement = engagement_modules.filter(
+        # Look for any of these engagement indicators:
+        # 1. Grade is not null (user submitted something)
+        # 2. State is not empty (user interacted with content)
+        # 3. Done status indicates completion
+        models.Q(grade__isnull=False) |
+        models.Q(state__isnull=False, state__gt='') |
+        models.Q(done__in=['f', 'i'])  # 'f' = finished, 'i' = incomplete but started
+    ).exists()
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: Same as above

File: lms/djangoapps/discussion/rest_api/tests/test_api.py [hunk 9]
```diff
@@ -51,6 +37,10 @@
     Role
 )
 from openedx.core.lib.exceptions import CourseNotFoundError, PageNotFoundError
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase, SharedModuleStoreTestCase
+from xmodule.modulestore.tests.factories import BlockFactory, CourseFactory
+from xmodule.partitions.partitions import Group, UserPartition
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_api.py]: Can you please check if changes in this file are needed or not as this file contains nothing but sorting of imports, if these changes are not needed can we revert these ? 

Thanks.

File: lms/djangoapps/discussion/rest_api/tests/test_api_v2.py [hunk 10]
```diff
@@ -24,31 +23,16 @@
 from pytz import UTC
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_api_v2.py:24]: Same as above, 

Can we only apply the changes that are needed and not the import sorting ? 

You can add imports that are required, you can remove imports that are not required, but don't sort them in these commits. That can be taken up separately. 

If its required to sort the imports for test cases to pass, please mention the reason in detail.

File: lms/djangoapps/discussion/rest_api/tests/test_discussions_notifications.py [hunk 11]
```diff
@@ -7,8 +7,10 @@
 
 import pytest
 
-from lms.djangoapps.discussion.rest_api.discussions_notifications import DiscussionNotificationSender, \
+from lms.djangoapps.discussion.rest_api.discussions_notifications import (
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_discussions_notifications.py]: Same as above we are just sorting the imports, that can be taken up separately.

File: lms/djangoapps/discussion/rest_api/tests/test_forms.py [hunk 12]
```diff
@@ -11,11 +11,7 @@
 from django.http import QueryDict
 from opaque_keys.edx.locator import CourseLocator
 
-from lms.djangoapps.discussion.rest_api.forms import (
-    UserCommentListGetForm,
-    CommentListGetForm,
-    ThreadListGetForm,
-)
+from lms.djangoapps.discussion.rest_api.forms import CommentListGetForm, ThreadListGetForm, UserCommentListGetForm
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_forms.py]: Same as above.

File: lms/djangoapps/discussion/rest_api/tests/test_permissions.py [hunk 13]
```diff
@@ -18,6 +16,8 @@
 from openedx.core.djangoapps.django_comment_common.comment_client.comment import Comment
 from openedx.core.djangoapps.django_comment_common.comment_client.thread import Thread
 from openedx.core.djangoapps.django_comment_common.comment_client.user import User
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_permissions.py]: Same as above.

File: lms/djangoapps/discussion/rest_api/api.py [hunk 14]
```diff
@@ -7,11 +7,9 @@
 import re
 from collections import defaultdict
 from datetime import datetime
-
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/api.py:10]: keep this change as is

File: lms/djangoapps/discussion/rest_api/api.py [hunk 15]
```diff
@@ -7,11 +7,9 @@
...
 from enum import Enum
 from typing import Dict, Iterable, List, Literal, Optional, Set, Tuple
 from urllib.parse import urlencode, urlunparse
-from pytz import UTC
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/api.py:14]: don't sort the import if it's not needed for your change

File: lms/djangoapps/discussion/rest_api/api.py [hunk 16]
```diff
@@ -66,17 +57,17 @@
 from openedx.core.djangoapps.django_comment_common.signals import (
     comment_created,
     comment_deleted,
-    comment_endorsed,
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/api.py:69]: Same as above.

File: lms/djangoapps/discussion/rest_api/api.py [hunk 17]
```diff
@@ -1636,6 +1630,8 @@ def update_comment(request, comment_id, update_data):
     if set(update_data) - set(actions_form.fields):
         serializer.save()
         comment_edited.send(sender=None, user=request.user, post=cc_comment)
+        # Create a new serializer with the updated comment to get fresh serialized data
+        serializer = CommentSerializer(cc_comment, context=context)
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/api.py]: Why do we need to recreated the serializer ?

File: lms/djangoapps/discussion/rest_api/discussions_notifications.py [hunk 18]
```diff
@@ -1,28 +1,26 @@
 """
 Discussion notifications sender util.
 """
-import re
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/discussions_notifications.py:4]: Just imports related changes, can be reverted

File: lms/djangoapps/discussion/rest_api/permissions.py [hunk 19]
```diff
@@ -7,19 +7,15 @@
...
-from common.djangoapps.student.roles import (
-    CourseInstructorRole,
-    CourseStaffRole,
-    GlobalStaff,
-)
-from lms.djangoapps.discussion.django_comment_client.utils import (
-    get_user_role_names,
-    has_discussion_privileges,
-)
+from common.djangoapps.student.roles import CourseInstructorRole, CourseStaffRole, GlobalStaff
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/permissions.py]: Same as above

File: lms/djangoapps/discussion/rest_api/serializers.py (hunk 2 above)
Comment (jcapphelix) [reply lms/djangoapps/discussion/rest_api/serializers.py]: @naincy128  this is not yet answered

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 20]
```diff
@@ -508,23 +567,27 @@ def create(self, validated_data):
         return thread
 
     def update(self, instance, validated_data):
+        # Extract reason codes for tracking events
+        edit_reason_code = validated_data.get('edit_reason_code')
+        close_reason_code = validated_data.get('close_reason_code')
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: @naincy128  this comment does seem valid from co-pilot, any reason we are not implementing it ?

File: lms/djangoapps/discussion/rest_api/urls.py [hunk 21]
```diff
@@ -22,7 +22,7 @@
     ReplaceUsernamesView,
     RetireUserView,
     ThreadViewSet,
-    UploadFileView,
+    UploadFileView
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/urls.py]: change only in imports, can be reverted

File: lms/djangoapps/discussion/rest_api/serializers.py (hunk 1 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/serializers.py]: Thanks! The extra import has been removed.

File: lms/djangoapps/discussion/rest_api/tests/test_api_v2.py [hunk 22]
```diff
@@ -2860,6 +2867,8 @@ def setUp(self):
         self.request.user = self.user
         CourseEnrollmentFactory.create(user=self.user, course_id=self.course.id)
         self.author = UserFactory.create()
+        self.register_get_user_response(self.author)
+        CourseEnrollmentFactory.create(user=self.author, course_id=self.course.id)
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/tests/test_api_v2.py]: Can you please check why this is needed ?

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 23]
```diff
@@ -496,3 +498,151 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+    if not is_enrolled:
+        return False
+
+    # Check if user has staff/moderator privileges
+    # Users with special roles are not considered regular learners
+    if _is_privileged_user(user, course_id):
+        return False
+
+    # User is a regular learner if they have meaningful engagement
+    return _check_user_engagement(user, course_id)
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: We don't need 2 cases, if a learner is not new then automatically he becomes a regular user. 

Also try to be consistent with naming schemes. Either we go with all the learners or we with all the users.

**Review by jcapphelix — CHANGES_REQUESTED** [review]

File: lms/djangoapps/discussion/rest_api/utils.py (hunk 23 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/utils.py]: Used one helper for both cases to remove duplication.

File: lms/djangoapps/discussion/rest_api/utils.py (hunk 5 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/utils.py]: Thanks for pointing that out —updated to use existing progress/completion models (BlockCompletion, PersistentSubsectionGrade) instead of direct heavy queries.

File: lms/djangoapps/discussion/rest_api/utils.py (hunk 6 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/utils.py]: Revised to use existing completion and grade models instead of costly direct queries.

File: lms/djangoapps/discussion/rest_api/utils.py (hunk 8 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/utils.py]: Revised to use existing completion and grade models instead of costly direct queries.

File: lms/djangoapps/discussion/rest_api/tests/test_api.py (hunk 9 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/tests/test_api.py]: Yes, we can revert this since it only contains import sorting and no functional changes.

File: lms/djangoapps/discussion/rest_api/tests/test_api_v2.py (hunk 22 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/tests/test_api_v2.py]: "self.user" needs enrollment to make API requests, and "self.author" needs enrollment to post threads/comments. Both are required for the test scenarios to be valid.

File: lms/djangoapps/discussion/rest_api/serializers.py (hunk 2 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/serializers.py]: Yes, updated to return "anonymous" for consistency.

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 24]
```diff
@@ -10,6 +10,7 @@
 
 from django.conf import settings
 from django.contrib.auth import get_user_model
+from django.contrib.auth.models import User  # lint-amnesty, pylint: disable=imported-auth-user
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: User model is already being fetched from Line number 12 and Line number 54, so we may not need this.

File: lms/djangoapps/discussion/rest_api/serializers.py [hunk 25]
```diff
@@ -275,6 +280,34 @@ def get_author_label(self, obj):
...
+
+        return get_user_learner_status(user, course.id)
+
+    def get_is_new_learner(self, obj):
+        """Return True if the author is a new learner."""
+        return self._get_learner_status_for_obj(obj) == "new"
+
+    def get_is_regular_learner(self, obj):
+        """Return True if the author is a regular learner."""
+        return self._get_learner_status_for_obj(obj) == "regular"
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/serializers.py]: Do we need 2 methods here ? 

Can we not determine if a learner is new or not with just one ? 

`is_new_learner=True`  would mean a new learner and regular if otherwise.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 26]
```diff
@@ -496,3 +498,117 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+    Args:
+        user: User object to check
+        course_id: Course key to check engagement in
+
+    Returns:
+        bool: True if user has meaningful engagement, False otherwise
+    """
+    # Check completion events
+    if BlockCompletion.objects.filter(user=user, context_key=course_id).exists():
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: This will almost always send false. 

Because this checks for entire course.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 27]
```diff
@@ -496,3 +498,117 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+# Optional backward-compatible wrappers (so existing code doesn't break)
+def is_user_new_learner(user, course_id):
+    """Return True if user is identified as a new learner."""
+    return get_user_learner_status(user, course_id) == "new"
+
+
+def is_user_regular_learner(user, course_id):
+    """Return True if user is identified as a regular learner."""
+    return get_user_learner_status(user, course_id) == "regular"
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: Do we need 2 methods ?

**Review by jcapphelix — CHANGES_REQUESTED** [review]
@naincy128  have given some changes, can you please check ?

File: lms/djangoapps/discussion/rest_api/serializers.py (hunk 24 above)
Comment (naincy128) [reply lms/djangoapps/discussion/rest_api/serializers.py]: Yes, it's already fetched at lines 12 and 54, so not needed.

File: lms/djangoapps/discussion/rest_api/utils.py [hunk 28]
```diff
@@ -496,3 +498,117 @@ def get_captcha_site_key_by_platform(platform: str) -> str | None:
...
+
+    # Enrollment check
+    is_enrolled = CourseEnrollment.objects.filter(
+        user=user,
+        course_id=course_id,
+        is_active=True
+    ).exists()
+
+    if not is_enrolled:
+        return "unenrolled"
```
Comment (jcapphelix) [inline lms/djangoapps/discussion/rest_api/utils.py]: @naincy128 Unenrolled users should not reach this code, so we can remove this portion and any other related logic.

**Review by jcapphelix — APPROVED** [review]
